# Handle report generation and plotting.  No real processing should happen here.

# Tables ------------------------------------------------------------------

#' Create genotype summary table
#'
#' Report the genotypes present in a processed dataset, converting sequences to
#' short names with optional custom naming.  If custom names are not given or
#' names for sequences are not present in the custom names, a name is
#' autogenerated from the sequence length and a truncated hash.
#'
#' @param tbl data frame as produced by \code{summarize_genotypes}.
#' @param allele.names optional data frame of custom allele names.  If present
#'   the columns "Sequence" and "Name" will be used to map exact sequences to
#'   custom names.
#' @param locus_attrs optional locus attributes data frame as produced by
#'   \code{load_locus_attrs}.  Used to order locus columns in the returned data
#'   frame.
#' @param hash.len number of characters of the sequence has to include when
#'   auto-generating allele names.  If set to zero the hash suffix is not
#'   included.
#'
#' @return data frame showing summary of genotypes.
#'
#' @export
report_genotypes <- function(tbl,
                             allele.names=NULL,
                             locus_attrs=NULL,
                             hash.len=6) {
  # At this point there are columns for sample, replicate, and then the loci.
  # If locus_attrs was given reorder the loci columns correspondingly.
  if (!missing(locus_attrs)) {
    locus_order <- order(match(colnames(tbl)[-(1:2)], rownames(locus_attrs)))
    tbl <- tbl[, c(1:2, 2+locus_order)]
  }
  # Name each entry in the table with either a custom or auto-generated short
  # name.
  for (j in 3:ncol(tbl)) {
    for (i in 1:nrow(tbl)) {
      if(is.na(tbl[i, j]))
        next
      row <- match(tbl[i, j], allele.names[, "Seq"])
      n <- as.character(allele.names[row, "Name"])
      if (is.na(n) || length(n) == 0)
        n <- make_allele_name(tbl[i, j], hash.len)
      tbl[i, j] <- n
    }
  }
  # If we have no replicates drop that column
  if (all(is.na(tbl$Replicate)))
    tbl <- tbl[, -2]
  # Blank out NA values
  tbl[is.na(tbl)] <- ''
  tbl
}

report_attribute <- function(dataset.summary, attrib, ...) {
  tbl <- summarize_attribute(dataset.summary, attrib, ...)
  # If we have no replicates drop that column
  if (all(is.na(tbl$Replicate)))
    tbl <- tbl[, -2]
  tbl
}

report_idents <- function(results, closest, hash.len) {
  tbl <- summarize_genotypes(results$summary)
  tbl.known <- summarize_genotypes_known(results$genotypes.known, tbl)
  # Now tbl.closest contains a full genotype table of those entries (possibly
  # multiple per sample) that are closest to each sample.  Next, interleave with
  # the samples themselves and group each set in a report table.
  tbl.closest <- do.call(rbind, lapply(names(closest), function(nm) {
    cbind(tbl.known[names(closest[[nm]]), ],
          Distance=closest[[nm]],
          Reference=nm)
  }))
  # I should clean up these functions to avoid this sort of mess.
  tbl.closest$Name <- tbl.closest$Sample
  tbl.closest$Sample <- tbl.closest$Reference
  tbl.closest <- tbl.closest[, -match("Reference", colnames(tbl.closest))]
  rownames(tbl.closest) <- paste(tbl.closest$Sample, tbl.closest$Name, sep='.')

  tbl.ident <- report_genotypes(tbl.closest[, -match(c("Distance", "Name"), colnames(tbl.closest))], hash.len = hash.len)
  tbl.ident <- cbind(tbl.ident, tbl.closest[, c("Distance", "Name")])

  tbl.obs <- report_genotypes(tbl, hash.len = hash.len)

  # Set common columns and combine tables
  tbl.obs$Distance <- NA
  tbl.obs$Name <- ""
  if( "Replicate" %in% colnames(tbl.obs))
    tbl.ident$Replicate <- ""
  tbl.combo <- rbind(tbl.obs, tbl.ident)

  # Order by sample by distance, with observations first, before known genotypes
  idx <- order(tbl.combo$Sample,
               tbl.combo$Distance,
               na.last = FALSE)
  tbl.combo <- tbl.combo[idx, ]
  tbl.combo$Distance[is.na(tbl.combo$Distance)] <- ""
  return(tbl.combo)
}

# Plots -------------------------------------------------------------------

#' Plot basic histogram of STR sample
#'
#' Given a processed STR sample, plot a histogram of counts per sequence length.
#'
#' @param samp data frame of dereplicated sequences.
#' @param main title of the plot.
#' @param locus.name name of the locus to match alleles for.  If unspecified the
#'   locus with the highest matched counts will be used.
#' @param sample.summary summary data frame as prepared by
#'   \code{summarize_sample}
#' @param cutoff_fraction numeric threshold for the fraction of locus-matching
#'   counts needed to call an allele.
#' @param xlim numeric range for x-axis.
#'
#' @export
#' @importFrom magrittr "%>%"
histogram <- function(samp,
                      main="Number of Reads by Sequence Length",
                      locus.name=NULL,
                      sample.summary=NULL,
                      cutoff_fraction=0.05,
                      xlim=range(samp$Length)) {

  ## Define colors and other plot parameters
  col.unlabeled   <- "#000000" # The non-locus sequences
  # Locus-labeled and matching all allele conditions
  col.filtered    <- "#FFAAAA"
  border.filtered <- "#990000"
  # Sequences matching alleles in sample.summary
  col.allele      <- "#FF0000"
  border.allele   <- "#FF0000"
  # Line showing threshold for allele calls
  col.cutoff <- rgb(0, 0, 0, 0.5)
  # Shaded region for filtered sequences
  col.region <- rgb(0, 0, 0, 30, maxColorValue = 255)
  lwd <- 5

  ## If there were no sequences at all, don't bother plotting.
  if (nrow(samp) == 0)
    return(NULL)

  ## Group sequences by length
  heights <- with(samp, {
    samp %>%
    dplyr::group_by(Length) %>%
    dplyr::summarize(TotalCount = sum(Count))
  })
  ymax <- max(heights$TotalCount)

  # Plot bars for all counts
  xlim <- range(heights$Length)
  ylim <- range(heights$TotalCount)
  plot(heights$Length,
       heights$TotalCount,
       type = "h",
       xlim = xlim,
       ylim = ylim,
       main = main,
       col = col.unlabeled,
       xlab = "Length (nt)",
       ylab = "Sequence Count",
       lwd = lwd,
       lend = 1)

  # If no locus name was given, take whatever locus showed the highest counts.
  if (missing(locus.name)) {
    cts <- with(samp, {
      subset(samp, MatchingLocus != "") %>%
        dplyr::group_by(MatchingLocus) %>%
        dplyr::summarize(Count=sum(Count))
    })
    cts <- cts[order(cts$Count, decreasing = T), ]
    locus.name <- if (nrow(cts)>1) {
      as.character(cts[[1, 'MatchingLocus']])
    } else {
      NA
    }
  }

  # Define subset of sample data that matches all locus conditions
  samp.filt <- with(samp,
                 subset(samp, MatchingLocus == locus.name &
                              MotifMatch &
                              LengthMatch))

  # Just the matching sequences, if there are any.
  if (nrow(samp.filt) > 0) {
    heights.filt <- with(samp.filt,{
      samp.filt %>%
        dplyr::group_by(Length) %>%
        dplyr::summarize(TotalCount = sum(Count))
    })
    points(heights.filt$Length,
           heights.filt$TotalCount,
           type = "h",
           col = col.filtered,
           lwd = lwd,
           lend = 1)
    # Shade the domain of the filtered data in gray
    xlim.filt <- range(samp.filt$Length)
    polygon(x = rep(xlim.filt, each=2),
            y = c(0, ymax, ymax, 0),
            col = col.region,
            border = NA)
    # Draw a line to mark the cutoff value for what's considered a prominent
    # count
    cutoff <- cutoff_fraction*sum(samp.filt$Count)
    abline(h = cutoff, col = col.cutoff)
  }

  # Draw bars for the exact allele sequences identified
  if (!is.null(sample.summary)) {
    pts.x <- c(sample.summary$Allele1Length, sample.summary$Allele2Length)
    pts.y <- c(sample.summary$Allele1Count, sample.summary$Allele2Count)
    pts.x <- pts.x[!is.na(pts.x)]
    pts.y <- pts.y[!is.na(pts.y)]
    if (length(pts.x) > 0)
      points(pts.x,
             pts.y,
             type = "h",
             col = col.allele,
             lwd = lwd,
             lend = 1)
  }

  # Legend
  legend(x='topright', bty='n',
         legend=c("Original",
                  "Filtered",
                  "Called Alleles",
                  "Unique Seq. Threshold"),
         col=c(col.unlabeled, col.filtered, col.allele, col.cutoff),
         pch=c(15, 15, 15, NA),
         lty=c(NA, NA, NA, 1))
}

#' Plot advanced histogram of STR sample
#'
#' Given a processed STR sample, plot a histogram of counts per unique sequence.
#' This is a more complicated version of \code{histogram}.
#'
#' @param samp data frame of dereplicated sequences.
#' @param stacked should bars be stacked together to add up to the total count
#'   for each length?  Otherwise they will be overlaid, so each bar height only
#'   corresponds to the count for that exact sequence.
#' @param main title of the plot.
#' @param locus.name name of the locus to match alleles for.  If unspecified the
#'   locus with the highest matched counts will be used.
#' @param sample.summary summary data frame as prepared by
#'   \code{summarize_sample}
#'
#' @export
#' @importFrom magrittr "%>%"
histogram2 <- function(samp,
                      stacked=TRUE,
                      main=NULL,
                      locus.name=NULL,
                      sample.summary=NULL) {
  col.unlabeled   <- "#DDDDDD" # The non-locus sequences
  # Locus-labeled sequences
  col.labeled     <- "#999999"
  border.labled   <- "#000000"
  # Locus-labeled and matching all allele conditions
  #col.filtered    <- "#FF6666"
  col.filtered    <- "#FFDDDD"
  border.filtered <- "#990000"
  # Sequences matching alleles in sample.summary
  col.allele      <- "#FF0000"
  border.allele   <- "#FF0000"
  lwd <- 5

  if (missing(locus.name)) {
    cts <- with(samp, {
      subset(samp, MatchingLocus != "") %>%
        dplyr::group_by(MatchingLocus) %>%
        dplyr::summarize(Count=sum(Count))
    })
    cts <- cts[order(cts$Count, decreasing = T), ]
    locus.name <- as.character(cts[[1, 'MatchingLocus']])
  }

  # TODO:
  #   support for showing stutter
  #   locus-specific options?
  #   un-stacked version
  #      show threshold

  if (stacked) {
    if (missing(main))
      main <- "Histogram of Sequence Lengths"
    # total counts at each length
    heights <- with(samp, {
        samp %>%
        dplyr::group_by(Length) %>%
        dplyr::summarize(TotalCount = sum(Count))
      })
    # Total ranges for stacked counts
    xlim <- range(heights$Length)
    ylim <- range(heights$TotalCount)
    heights.loci <- with(samp, {
      samp %>%
      subset(MatchingLocus != "") %>%
      dplyr::group_by(Length) %>%
      dplyr::summarize(TotalCount = sum(Count))
    })

    # Plot histogram of total counts, gray in background
    plot(heights$Length,
         heights$TotalCount,
         type = "h",
         xlim = xlim,
         ylim = ylim,
         main = main,
         col = col.unlabeled,
         xlab = "Length (nt)",
         ylab = "Sequence Count",
         lwd = lwd,
         lend = 1)

    # Stack counts for individual sequences for the locus-labeled cases.  If any
    # rows match specific conditions, color those rectangles separately.
    for (len in unique(samp$Length)) {
      idx <- with(samp, Length == len & MatchingLocus != "")
      chunk <- samp[idx, ]
      if (nrow(chunk) < 1)
        next
      chunk <- chunk[order(chunk$Count), ]
      chunk$Cumsum <- cumsum(chunk$Count)
      for (n in 1:nrow(chunk)) {
        x <- c(len - lwd / 10,
               len - lwd / 10,
               len + lwd / 10,
               len + lwd / 10)
        y <- c(chunk[n, "Cumsum"] - chunk[n, "Count"],
               chunk[n, "Cumsum"],
               chunk[n, "Cumsum"],
               chunk[n, "Cumsum"] - chunk[n, "Count"])
        # Is the sequence matching all criteria for a candidate allele?
        am <- allele_match(chunk[n, ], locus.name)
        if(is.na(am)) am <- F
        # Is the sequence an exact match for an identified allele?
        is_allele <- chunk[n, 'Seq'] %in%
          as.character(unlist(sample.summary[c('Allele1Seq', 'Allele2Seq')]))
        # Color each rectangle according to its category as defined by the
        # above.  The default will be the locus-labeled colors.
        col <- col.labeled
        col.border <- border.labled
        if (is_allele) {
          col <- col.allele
          col.border <- border.allele
        } else if (am) {
          col <- col.filtered
          col.border <- border.filtered
        }
        polygon(x, y, col = col, border = col.border, lwd=0.5)
      }
    }
  }
}

# if include.blanks = TRUE then we should have a full count of all alleles when
# adding up the numbers on the extra axis labels.
plot_alignment <- function(alignment, labels=NULL, include.blanks=FALSE, ...) {
  # Convert to character and remove blanks if specified
  if (is.character(alignment))
    seqs <- alignment
  else
    seqs <- as.character(alignment)
  if (! include.blanks)
    seqs <- seqs[grep('^-+$', seqs, invert = TRUE)]
  # Create grouping factor using sequence length (just strip out the gap
  # character to get the original length back).
  groups <- factor(paste('  ', nchar(gsub('-', '', seqs)), 'bp'))
  par(mar = c(5, 5, 4, 5))
  dnaplotr::plotDNA(seqs,
                    groups = groups,
                    ...)
  # Add faint lines between all sequences
  for(i in seq_along(seqs))
    abline(h=i+0.5, col=rgb(0.5, 0.5, 0.5, 0.5))
  # Add a label for every unique sequence, using the names of the supplied
  # sequences
  if (missing(labels))
    labels <- sapply(strsplit(names(seqs), '_'), '[', 2)
  axis(4,
       at = 1:length(seqs),
       labels = labels,
       tick = F,
       padj = -2.5,
       cex.axis=0.6)
}

# Distance Matrices -------------------------------------------------------


# A skewed 0 -> 1 scale for color-coding distance tables
make.dist_scale <- function(n) {
  ((0:n)/n)**(1/3)
}

plot_dist_mat <- function(dist_mat, num.alleles=max(dist_mat),
                          dist.display_thresh=round(num.alleles*2/3),
                          ...) {
  labels <- matrix(character(length(dist_mat)), nrow=nrow(dist_mat))
  idx <- dist_mat <= dist.display_thresh
  labels[idx] <- dist_mat[idx]
  diag(labels) <- ''
  dist_scale <- make.dist_scale(num.alleles)
  color <- rgb(red=1, green=dist_scale, blue=dist_scale)

  vals <- dist_mat
  pheatmap::pheatmap(vals,
           color = color,
           display_numbers = labels,
           treeheight_row = 0,
           breaks = 0:num.alleles,
           clustering_distance_rows = as.dist(dist_mat),
           clustering_distance_cols = as.dist(dist_mat),
           ...)
}


# Heatmaps ----------------------------------------------------------------

#' Render heatmap of STR attribute across samples and loci
#'
#' Given a cross-sample summary data frame as produced by \code{analyze_dataset}
#' and the name of a column (e.g., Stutter, Homozygous, ProminentSequences),
#' plot a heatmap of the values for that attribute, with sample identifiers on
#' rows and loci on columns.  The attribute will be coerced to numeric.
#'
#' @param results_summary cross-sample summary data frame as produced by
#'   \code{analyze_dataset}.
#' @param attribute character name of column in results_summary to use.
#' @param label.by vector of column names to use when writing the genotype
#'   summary values on top of the heatmap cells (default: allele lengths).
#' @param color vector of colors passed to \code{pheatmap}.
#' @param breaks vector of breakpoints passed to \code{pheatmap}.
#' @param ... additional arguments to \code{pheatmap}.
#'
#' @export
plot_heatmap <- function(results_summary,
                         attribute,
                         label.by = c('Allele1Length', 'Allele2Length'),
                         color=c("white", "pink"),
                         breaks=NA,
                         ...) {
  tbl <- summarize_attribute(results_summary, attribute)
  data <- tbl[, -(1:2)] + 0
  tbl.labels <- summarize_genotypes(results_summary, vars = label.by)
  labels <- tbl.labels[, -(1:2)]
  data[is.na(labels)] <- NA
  labels[is.na(labels)] <- ''

  if (min(data,na.rm=T) == max(data, na.rm=T))
    breaks <- c(max(data))
  pheatmap::pheatmap(data,
                     cluster_rows = F,
                     cluster_cols = F,
                     display_numbers = labels,
                     breaks = breaks,
                     color = color,
                     ...)
}

#' Plot heatmap of suspected PCR stutter
#'
#' Given a cross-sample summary data frame as produced by
#' \code{analyze_dataset}, plot a heatmap showing which samples had alleles
#' ignored due to suspected PCR stutter, with sample identifiers on rows and
#' loci on columns.
#'
#' @param results_summary cross-sample summary data frame as produced by
#'   \code{analyze_dataset}.
#' @param ... additional arguments passed to \code{plot_heatmap}.
#'
#' @export
plot_heatmap_stutter <- function(results_summary, ...) {
  plot_heatmap(results_summary,
               "Stutter",
               legend = FALSE,
               ...)
}

#' Plot heatmap of homozygous samples
#'
#' Given a cross-sample summary data frame as produced by
#' \code{analyze_dataset}, plot a heatmap showing which samples appear
#' homozygous, with sample identifiers on rows and loci on columns.
#'
#' @param results_summary cross-sample summary data frame as produced by
#'   \code{analyze_dataset}.
#' @param ... additional arguments passed to \code{plot_heatmap}.
#'
#' @export
plot_heatmap_homozygous <- function(results_summary, ...) {
  plot_heatmap(results_summary,
               "Homozygous",
               legend = FALSE,
               ...)
}

#' Plot heatmap of samples with multiple prominent sequences
#'
#' Given a cross-sample summary data frame as produced by
#' \code{analyze_dataset}, plot a heatmap showing samples with more than two
#' prominent sequences in their analysis output, with sample identifiers on rows
#' and loci on columns.
#'
#' @param results_summary cross-sample summary data frame as produced by
#'   \code{analyze_dataset}.
#' @param ... additional arguments passed to \code{plot_heatmap}.
#'
#' @export
plot_heatmap_prominent_seqs <- function(results_summary, ...) {
  # Create a color ramp going from white for 0, 1, or 2 prominent seqs, and
  # shades of red for higher numbers.
  color_func <- colorRampPalette(c("white", "red"))
  ps <- results_summary[!is.na(results_summary$Allele1Seq), "ProminentSeqs"]
  # Deep red will only be used if somehow there are a whole lot of extra
  # sequences (say, 8); otherwise it should just go up to a pink color.
  colors <- color_func(max(8, max(ps)+1))
  # Stay white for 0 - 2
  colors[1:3] <- rep(colors[1], 3)
  # Truncate to actual number of peaks
  colors <- colors[1:max(ps) + 1]
  plot_heatmap(results_summary,
               "ProminentSeqs",
               color = colors,
               breaks = 0:(length(colors)),
               ...)
}

#' Plot heatmap of proportion of allele sequence counts
#'
#' Given a cross-sample summary data frame as produced by
#' \code{analyze_dataset}, plot a heatmap showing the proportion of matching
#' sequences for the identified alleles versus all matching sequences, with
#' sample identifiers on rows and loci on columns.
#'
#' @param results_summary cross-sample summary data frame as produced by
#'   \code{analyze_dataset}.
#' @param ... additional arguments passed to \code{plot_heatmap}.
#'
#' @export
plot_heatmap_proportions <- function(results_summary, ...) {
  cts <- results_summary[, c('Allele1Count', 'Allele2Count')]
  prop.counted <- rowSums(cts, na.rm=T)/results_summary$CountLocus
  results_summary$ProportionCounted <- prop.counted
  color_func <- colorRampPalette(c("red", "white"))
  breaks <- seq(0, 1, 0.001)
  colors <- color_func(length(breaks) - 1)
  plot_heatmap(results_summary,
               'ProportionCounted',
               color = colors,
               breaks = breaks,
               ...)
}


# Markdown ----------------------------------------------------------------

# k: existing kable-generated table
# grouping: logical vector with TRUE for rows to start groups with,
# optionally named
k_group_rows <- function(k, grouping) {
  idx <- which(grouping)
  labels <- names(grouping)[idx]
  if (is.null(labels))
    labels <- seq_along(idx)
  for(i in seq_along(idx)) {
    start_row <- idx[i]
    end_row <- idx[i+1] - 1
    if (is.na(end_row))
      end_row <- length(grouping)
    k <- kableExtra::group_rows(k, labels[i], start_row, end_row)
  }
  k
}

k_row_spec <- function(k, idx.rows, ...) {
  for (idx in idx.rows) {
    k <- kableExtra::row_spec(k, idx, ...)
  }
  k
}

# convenience function for post-processing report_genotypes() output
kable_genotypes <- function(data) {
  bootstrap_options <- c("striped", "hover", "condensed")
  k <- knitr::kable(data, row.names = FALSE, format = "html")
  kableExtra::kable_styling(k,
                            bootstrap_options = bootstrap_options,
                            full_width = F)
}

# Write markdown tables to standard output for report_genotypes()
rmd_kable_genotypes <- function(results, locus_attrs, hash.len,
                                locus_chunks=NULL) {
  tbl.g <- summarize_genotypes(results$summary)
  tbl <- report_genotypes(tbl.g,
                          locus_attrs = locus_attrs,
                          hash.len = hash.len)
  if (!is.null(locus_chunks)) {
    prefix <- match(c("Sample", "Replicate"), colnames(tbl))
    prefix <- prefix[!is.na(prefix)]
    for(chunk_name in names(locus_chunks)) {
      locus_cols <- paste(rep(locus_chunks[[chunk_name]], each=2),
                          c(1,2),
                          sep='_')
      m <- match(locus_cols, colnames(tbl))
      m <- m[!is.na(m)]
      cat(paste0("\n\n### Loci: ", chunk_name, "\n\n"))
      cat(kable_genotypes(tbl[, c(prefix, m)]))
    }
  } else {
    cat(kable_genotypes(tbl))
  }
}

kable_idents <- function(tbl, closest) {
  # Remove columns that will be represented in other ways (sample/remplicate in
  # row groupings)
  idx.remove <- match(c("Sample", "Replicate"), colnames(tbl))
  idx.remove <- idx.remove[!is.na(idx.remove)]
  tbl <- tbl[, -idx.remove]

  # Create basic table
  bootstrap_options <- c("hover", "condensed")
  k <- knitr::kable(tbl, row.names = FALSE, format = "html")
  k <- kableExtra::kable_styling(k,
                                 bootstrap_options = bootstrap_options,
                                 full_width = F)

  # Group rows by sample
  obs.select <- tbl$Distance == ""
  names(obs.select) <- paste("Sample", rownames(tbl))
  k <- k_group_rows(k, obs.select)

  # Bold rows containing a single identification per sample
  # (find the original samples, and then go one farther)
  ids <- names(closest[sapply(closest, function(x) length(x) == 1)])
  idx.rows <- match(ids, rownames(tbl)) + 1
  k <- k_row_spec(k, idx.rows, bold = TRUE)

  k
}

rmd_kable_idents <- function(results, hash.len, range, maximum, locus_chunks=NULL) {
  closest <- find_closest_matches(results$dist_mat_known,
                                  range = range,
                                  maximum = maximum)
  tbl.combo <- report_idents(results, closest, hash.len)
  kable_idents(tbl.combo, closest)
}
